order/conditionally_complete_lattice.lean

```diff
@@ -0,0 +1,469 @@
+/-
+Copyright (c) 2018  Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+Adapted from the corresponding theory for complete lattices.
+
+Theory of conditionally complete lattices.
+
+A conditionally complete lattice is a lattice in which every non-empty bounded subset s
+has a least upper bound and a greatest lower bound, denoted below by Sup s and Inf s.
+Typical examples are real, nat, int with their usual orders.
+
+The theory is very comparable to the theory of complete lattices, except that suitable
+boundedness and non-emptyness assumptions have to be added to most statements.
+We introduce two predicates bdd_above and bdd_below to express this boundedness, prove
+their basic properties, and then go on to prove most useful properties of Sup and Inf
+in conditionally complete lattices.
+
+To differentiate the statements between complete lattices and conditionally complete
+lattices, we prefix Inf and Sup in the statements by c, giving cInf and cSup. For instance,
+Inf_le is a statement in complete lattices ensuring Inf s ≤ x, while cInf_le is the same
+statement in conditionally complete lattices with an additional assumption that s is
+bounded below.
+-/
+import order.lattice order.complete_lattice tactic.finish data.set.countable
+
+set_option old_structure_cmd true
+
+open preorder set lattice
+
+universes u v
+variables {α : Type u} {β : Type v}
+
+section preorder
+variables [preorder α] {s t: set α} {a b: α}
+
+/-Sets bounded above and bounded below.-/
+def bdd_above (s: set α) := (∃x, ∀y∈s, y ≤ x)
+def bdd_below (s: set α) := (∃x, ∀y∈s, x ≤ y)
+
+/-Introduction rules for boundedness above and below-/
+lemma bdd_aboveI (a:α) (H: ∀y∈s, y≤a) : bdd_above s := ⟨a, H⟩
+lemma bdd_belowI (a:α) (H: ∀y∈s, a≤y) : bdd_below s := ⟨a, H⟩
```

Comment:
Usually we name introduction rules `*.mk`. In this case, though, I'd skip these definitions entirely, and just use `⟨a, H⟩` when it comes up.

Reply:
I have a use case below, in a tactics proof. I could replace it by `refine ⟨w, _⟩` in this proof, but the intent is more clear when using the introduction rule, I think.